src/HydraAuctionOnchain/Validators/StandingBid.hs

```
@@ -0,0 +1,283 @@
+module HydraAuctionOnchain.Validators.StandingBid
+  ( standingBidValidator
+  ) where
+
+import HydraAuctionOnchain.Helpers
+  ( pdecodeInlineDatum
+  , pfindUniqueInputWithToken
+  , pfindUniqueOutputWithAddress
+  , putxoAddress
+  )
+import HydraAuctionOnchain.MintingPolicies.Auction
+  ( auctionEscrowTokenName
+  , standingBidTokenName
+  )
+import HydraAuctionOnchain.Types.AuctionTerms (PAuctionTerms, pbiddingPeriod)
+import HydraAuctionOnchain.Types.Error (ToErrorCode (toErrorCode), err, perrMaybe)
+import HydraAuctionOnchain.Types.StandingBidState (PStandingBidState, pvalidateNewBid)
+import HydraAuctionOnchain.Validators.AuctionEscrow (pisConcluding)
+import Plutarch.Api.V1.Value (pvalueOf)
+import Plutarch.Api.V2
+  ( PCurrencySymbol
+  , PScriptContext
+  , PScriptPurpose (PSpending)
+  , PTxInInfo
+  , PTxInfo
+  , PTxOut
+  )
+import Plutarch.Extra.Interval (pcontains)
+import Plutarch.Extra.Maybe (pmaybe)
+import Plutarch.Extra.ScriptContext (ptryFromRedeemer, ptryOwnInput, ptxSignedBy)
+import Plutarch.Monadic qualified as P
+
+--------------------------------------------------------------------------------
+-- Redeemers
+--------------------------------------------------------------------------------
+
+data PStandingBidRedeemer (s :: S)
+  = NewBidRedeemer (Term s (PDataRecord '[]))
+  | MoveToHydraRedeemer (Term s (PDataRecord '[]))
+  | ConcludeAuctionRedeemer (Term s (PDataRecord '[]))
+  deriving stock (Generic)
+  deriving anyclass (PlutusType, PIsData, PShow, PEq)
+
+instance DerivePlutusType PStandingBidRedeemer where
+  type DPTStrat _ = PlutusTypeData
+
+--------------------------------------------------------------------------------
+-- Errors
+--------------------------------------------------------------------------------
+
+data PStandingBidError (s :: S)
+  = StandingBid'Error'OwnInputMissingToken
+  | StandingBid'Error'UnexpectedTokensMintedBurned
```

I think that you should also throw an error if there is more than one input from the standing bid validator.

I apply this "too many script inputs" check in every validator, so that I don't have to worry about double-satisfaction problems. Later on, when we optimize performance, we can selectively remove this check where having inputs from multiple auctions is known not to cause problems.